core/src/main/java/com/digitalpebble/stormcrawler/filtering/URLFilter.java

```diff
@@ -17,6 +17,8 @@
 import com.digitalpebble.stormcrawler.Metadata;
 import com.digitalpebble.stormcrawler.util.Configurable;
 import java.net.URL;
+import org.jetbrains.annotations.NotNull;
```

Comment:
As far as i know there is no "official" annotation, so I'll just write why I used the Jetbrains one. (see [[1]](#sources) and [[2]](#sources))

There was once the javax.annotation thing, but as far as i know that one is basically dead. 

An other framework is SpotBugs the successor of FindBugs, but I have never used it because of trust issues (FindBugs was already abandoned, why shouldn't happen the same gain?) and there are more "standardly" feeling frameworks.

Another framework is Lombok, but this one is modifying the code to add real nullchecks to the code during runtime. But I don't like the idea of having a framework for annotating null-handling modifying written code, that only causes problems in the long run when the framework ceases to exist.

The next Framework is the Checker Framework. This is also the framework that Guava uses and can be seen as one of the more "standardly" feeling frameworks. But CF is quite big and does a lot of different checks that are far beyond simple documentation of "you can give me null" and properly using it seems quite bothersome because it has many annotations and most features are not supported by other checker frameworks and IDEs. (But I haven't used it so far, so I can only make assumptions about that support point. But I haven't found anything that says what is supported.)


The reasons why I chose `org.jetbrains.annotations` over the others are as follows:
- They only exist during compile time and are more part of the documentation like part of the code. So adding them like this is enough and we don't have to add any weird dependencies on compiler-plugins or whatnot:
```xml
<!-- https://mvnrepository.com/artifact/org.jetbrains/annotations -->
<dependency>
    <groupId>org.jetbrains</groupId>
    <artifactId>annotations</artifactId>
    <version>23.0.0</version>
    <scope>compile</scope>
</dependency>
```
- They are supported by CF, Eclipse, FindBugs, SpotBugs etc. (Regarding my research this annotations are basically supported by every big IDE and checking framework), giving them (at least for me) a very "standardly" feeling. (see [[5]](#sources) and [[6]](#sources) for an example how to use it in eclipse.)
- The amount of annotations is small (14) and are quite straight forward regarding their meaning [[3]](#sources).  Even if the framework has to be removed on a later date, there won't be much fuss about it because there are only 14 of them and you usually only use `@Nullable`, `@NotNull`, `@TestOnly` and `@Contract`.
- Some Apache Projects like Geode and Sling, and other opensource projects like testcontainer already use these annotations. Apache Spring explains it's usage of these annotations like this [[4]](#sources): 
    > The annotations used within Sling are based on the [Jetbrains Annotations](https://www.jetbrains.com/help/idea/nullable-and-notnull-annotations.html). Although introduced by the company that offers the IntelliJ IDEA IDE, those annotations are understood by most of the tools and used by other Apache Projects like Apache Oak.


## Sources
1. https://jcp.org/en/jsr/detail?id=305
2. An exhaustive list of nullness checking frameworks: https://checkerframework.org/manual/#nullness-related-work
3. https://github.com/JetBrains/java-annotations
4. https://sling.apache.org/documentation/development/null-analysis.html
5. https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Ftasks%2Ftask-using_external_null_annotations.htm
6. https://sling.apache.org/documentation/development/null-analysis.html#use-with-eclipse

Reply:
OK, you  convinced me